File: alphazero/cpp_impl/env.py
import copy
from enum import Enum

import numpy as np
import gym
import gym_minigrid


class Action(Enum):
    left = 0
    right = 1
    forward = 2


class EnvWrapper:
    def __init__(self, env, params, actions):
        self.env = env
        self.actions = actions
        self.params = params

    def step(self, action):
        action = self.actions[action]
        obs, reward, done, _ = self.env.step(action)
        return (
            np.append(self.env.agent_pos, self.env.agent_dir),
            reward ** self.params["reward_exponent"],
            done,
            None
        )

    def reset(self):
        self.env.reset()
        return np.append(self.env.agent_pos, self.env.agent_dir)


def prepare_minigrid(game, params):
    env = gym.make('MiniGrid-Empty-%s-v0' % game)
    actions = {
        0: Action.left,
        1: Action.right,
        2: Action.forward
    }
    env.actions = Action
    return EnvWrapper(env, params, actions)


def init(game, params):
    return prepare_minigrid(game, params)


def step(env, action):
    obs, reward, done, _ = env.step(action)
    return list(obs), reward, done


def reset(env):
    return list(env.reset())


def clone(env):
    return copy.deepcopy(env)

File: alphazero/cpp_impl/env.cpp
#include <cmath>
#include "env.hpp"
#include "cfg.hpp"
#include <iostream>

//
// TODO At some point this will need templating for all envs.
//

void
Env::init(std::string game, json params) {
  reward_exponent = params["reward_exponent"];

  std::set<std::pair<int, int>> blocks;
  int width = 3;
  if (game == "5x5")
    width = 5;
  if (game == "8x8")
    width = 8;
  if (game == "16x16")
    width = 16;
  for (int i = 0; i < width; ++i) {
    blocks.insert({0, i});
    blocks.insert({width - 1, i});
    blocks.insert({i, 0});
    blocks.insert({i, width - 1});
  }
  grid_world = GridWorldEnv(width, width, blocks);
}

std::tuple<std::vector<int>, double, bool>
Env::step(int action) {
  std::tuple<std::vector<int>, double, bool> ret = grid_world.step(action);
  double& reward = std::get<1>(ret);
  if (reward != 0)
    reward = std::pow(reward, reward_exponent);
  return ret;
}

std::vector<int>
Env::reset() {
  return grid_world.reset();
}

std::unique_ptr<Env>
Env::clone() {
  auto env = std::make_unique<Env>();
  // TODO Could be better, eg move these things into a constructor.
  env->grid_world = grid_world;
  env->reward_exponent = reward_exponent;
  return env;
}
